Migrate to paper-scroll-header-panel (closes #206 and #175)

chuckh opened the pull request:
Migrate from paper-header-panel to paper-scroll-header-panel. Both middle and bottom titles condense correctly.

Closes #206 and #175.

![psk-paper-scroll-header-panel](https://cloud.githubusercontent.com/assets/19915/8477215/548b6b3a-207d-11e5-8dc6-57f7f004772a.gif)

arthurvr commented on app/scripts/app.js:
```diff
@@ -30,6 +30,27 @@ subject to an additional IP rights grant found at http://polymer.github.io/PATEN
+    var appName = document.querySelector('.app-name');
+    var middleContainer = document.querySelector('.middle-container');
+    var bottomContainer = document.querySelector('.bottom-container');
+    var d = e.detail;
```
Let's keep things readable and don't use one-letter variable names ;)

chuckh replied:
@arthurvr does changing from
```
var d = e.detail;
var m = d.height - d.condensedHeight;
 var y = Math.min(1, d.y / m);
```
to
```
var detail = e.detail;
var heightDiff = detail.height - detail.condensedHeight;
var yRatio = Math.min(1, detail.y / heightDiff);
```
seem reasonable too you?

arthurvr replied:
:+1: Way more readable at least.

arthurvr commented on app/scripts/app.js:
```diff
@@ -30,6 +30,27 @@ subject to an additional IP rights grant found at http://polymer.github.io/PATEN
     // imports are loaded and elements have been registered
   });
 
+  // custom transformation: scale header's titles
```
We should be consistent with this. Do we start comments with a capital letter or not? It seems like the rest of the code actually does.

chuckh replied:
I will make them start with capital letters.

chuckh commented:
I made @arthurvr suggested changes to variable names and comments.

chuckh commented:
@addyosmani I added additional comments to help explain what is happening. Also added
variable maxMiddleScale to make it easier to change size of condensed appName.

kdzwinel commented on app/index.html:
```diff
@@ -84,11 +85,13 @@
+      </paper-scroll-header-panel>
+
+      <!-- Main Area -->
+      <paper-scroll-header-panel main condenses keep-condensed-header>
```
Closing tag is still `</paper-header-panel>` instead of `</paper-scroll-header-panel>`

chuckh replied:
@kdzwinel nice catch. I fixed this with the latest commit.

kdzwinel commented on app/scripts/app.js:
```diff
@@ -30,6 +30,31 @@ subject to an additional IP rights grant found at http://polymer.github.io/PATEN
+    Polymer.Base.transform('scale(' + scaleBottom + ') translateZ(0)', bottomContainer);
+
+    // Scale middleContainer appName
+    Polymer.Base.transform('scale(' + scaleMiddle + ') translateZ(0)', appName);
```
Simple `scale` won't unfortunately work if text breaks into two lines:

<img width="563" alt="screen shot 2015-07-07 at 01 56 59" src="https://cloud.githubusercontent.com/assets/985504/8535978/78c196f6-244b-11e5-8ba9-79230371d438.png">

chuckh replied:
@kdzwinel any suggestions for a better approach?

kdzwinel replied:
Maybe forcing single line (`white-space: nowrap; overflow: hidden; text-overflow: ellipsis;`) as soon as animation starts and then using `scale` would be a good compromise?

kdzwinel commented on app/scripts/app.js:
```diff
@@ -30,6 +30,31 @@ subject to an additional IP rights grant found at http://polymer.github.io/PATEN
+  addEventListener('paper-header-transform', function(e) {
+    var appName = document.querySelector('.app-name');
+    var middleContainer = document.querySelector('.middle-container');
+    var bottomContainer = document.querySelector('.bottom-container');
```
For better performance these three should probably be cached outside the listener (same as [here](view-source:https://elements.polymer-project.org/bower_components/paper-scroll-header-panel/demo/index.html)).

chuckh replied:
Having these three outside does not work with PSK. I put them inside to get them work.

kdzwinel replied:
Hm, that's strange. Try putting the whole thing inside of `window.addEventListener('WebComponentsReady', function() {` ([line 24](https://github.com/PolymerElements/polymer-starter-kit/pull/233/files#diff-685dfa901a119f6590e773891ffd6052R24)).

chuckh replied:
That worked as long as you are using Shady DOM, does not work with Shadow Dom. Current method works with both.

kdzwinel replied:
Didn't know that - thanks! So, maybe something like this?

    var appName = null;

    addEventListener('paper-header-transform', function(e) {
       appName = appName || document.querySelector('.app-name');

Not a beauty, but this listener is fired a lot, so I'd be great to optimize it a bit.

chuckh replied:
This worked for both Shady DOM and Shadow DOM @kdzwinel. @addyosmani and @arthurvr what do you think about this approach? 

```
  var appName = null;
  var middleContainer = null;
  var bottomContainer = null;
  addEventListener('paper-header-transform', function(e) {
    appName = appName || document.querySelector('.app-name');
    middleContainer = middleContainer || document.querySelector('.middle-container');
    bottomContainer = bottomContainer || document.querySelector('.bottom-container');
    var detail = e.detail;
    var heightDiff = detail.height - detail.condensedHeight;
    var yRatio = Math.min(1, detail.y / heightDiff);
    var maxMiddleScale = 0.50;  // appName max size when condensed. The smaller the number the smaller the condensed size.
    var scaleMiddle = Math.max(maxMiddleScale, (heightDiff - detail.y) / (heightDiff / (1-maxMiddleScale))  + maxMiddleScale);
    var scaleBottom = 1 - yRatio;

    // Move/translate middleContainer
    Polymer.Base.transform('translate3d(0,' + yRatio * 100 + '%,0)', middleContainer);

    // Scale bottomContainer and bottom sub title to nothing and back
    Polymer.Base.transform('scale(' + scaleBottom + ') translateZ(0)', bottomContainer);

    // Scale middleContainer appName
    Polymer.Base.transform('scale(' + scaleMiddle + ') translateZ(0)', appName);
  });
```

addyosmani commented:
Changes LGTM. As our direction here is starting to cause a little friction with some other incoming PRs unaware of these changes, I think now is a good time to land this in master. It's definitely good enough for the next release.
